docs/RFCS/data_protection/ssec.md

```diff
@@ -0,0 +1,104 @@
+
+
+## Motivation
+
+In order to improve the data protection feature and to provide LeoFS as an enterprise secure data storage, SSE-C must be implemented as one of the data protection functions. After we implement this function, we can provide LeoFS for [GDPR](https://en.wikipedia.org/wiki/General_Data_Protection_Regulation), [FHIR](https://en.wikipedia.org/wiki/Fast_Healthcare_Interoperability_Resources) and other data security needs.
+
+
+## Detailed design
+### Prerequisites
+
+The data path between LeoFS' clients and LeoGateway is securely protected by HTTPS and the internal communication between LeoGateway and LeoStorage is also protected by Erlang distribution over TLS [^2].
+
+LeoFS never store **keys of all customers** and the **original checksum** and **size** of an object on metadata. Since we suggest the requirement that *"Anyone who has access to the stored data can try to manipulate the data - even if the data is encrypted. (Quoted from 'Go implementation of the Data At Rest Encryption (DARE) format [^3])"*, storing the original checksum (MD5) and the size could probably allow attackers to derive the original content by brute force attack based on the size and checksum.
+
+
+### Requirements
+
+LeoStorage encrypts and decrypts objects because LeoStorage's CPU usage tends to be considerably lower than LeoGateway. Furthermore, in many cases, the number of LeoStorage's in a LeoFS storage system is larger than the number of LeoGateway and thus, load balancing can be expected.
+
+In each of the following operations, LeoStorage handles encryption and decryption of an object after an operation is requested from LeoGateway.
+
+* GET
+* Range GET
+* HEAD
+* PUT
+* Multipart Upload
+* COPY
+
+LeoGateway never caches objects to securely protect.
+
+
+### Request Handling on LeoGateway
+
+When LeoGateway receives an SSE-C request from a LeoFS' client, it retrieves the encryption key information from the request header, then it sends the information and the object *(only if PUT or Multipart Upload operation was requested)* to LeoStorage.
+
+When a requested object corresponds to Large Size Object, LeoGateway divides the object to plural chunks of it, then it sends each chunk with the encryption key information to LeoStorage in order to individually execute object encryption.
+
+
+#### Encryption key information in HTTP Header:
+
+| Name  | Description  |
+|----|----|
+|`x-amz-server-side​-encryption​-customer-algorithm`| Encryption algorithm *(AES 256)* |
+|`x-amz-server-side​-encryption​-customer-key` | 256-bit, base64-encoded encryption key |
+| `x-amz-server-side​-encryption​-customer-key-MD5` | Base64-encoded 128-bit MD5 digest of the encryption key |
+
+When LeoGateway accepts COPY operation, a client sends a request header which includes the following items. The destination object is encrypted by using those items.
```

Comment:
This sentence is not true.
The following headers can be used to decrypt the source object if that's encrypted using SSE-C according to the SSEC-C spec. That said, if the source object is not encrypted then those headers are not required.
```
If the source object is encrypted using SSE-C,
you must provide encryption key information using the following headers so that Amazon S3 can decrypt the object for copying.
```
If users want to encrypt the object on the destination then users would have to use the above three headers prefixed by `x-amz-server-side-encryption`.